test/unit/Container.spec.ts

```diff
@@ -128,13 +112,13 @@ describe("Container", function() {
             }
         }
 
-        function createCar (engine: Engine) {
-            return new Car(engine);
+        class CarFactory {
+            @Factory()
+            public static createCar (engine: Engine): Car {
```

Comment:
`@Factory` decorator should be removed to remove "double-behaviour" and remove cases when service is initialized without knowing where how and who affected service creation.

For your second question, function in factory is enough. I think there is no need to complicate things with such complex use cases.